From d6a6bee6162b663ba952319cd0391a2029020fd0 Mon Sep 17 00:00:00 2001
From: Sandeep Kunta <skunta@codeaurora.org>
Date: Tue, 28 Jul 2015 13:07:07 +0530
Subject: [PATCH] Fix delay in bringing incallui for MO/MT calls

The present change has below fixes.
1. For incoming calls bringup incallui immediately instead of
waiting for 500ms as in current implementation.
2. Avoid updating call state update of "CONNECTING" to InCallUI
to avoid redundant call state updates and there by reduce InCallUI
bringup delay

Change-Id: Id78b3eff63aade6c4610b014b734e227d5a600ad
CRs-Fixed: 874691
---

diff --git a/src/com/android/server/telecom/Call.java b/src/com/android/server/telecom/Call.java
index d6efdc8..de5296a 100644
--- a/src/com/android/server/telecom/Call.java
+++ b/src/com/android/server/telecom/Call.java
@@ -22,6 +22,7 @@
 import android.net.Uri;
 import android.os.Bundle;
 import android.os.Handler;
+import android.os.SystemProperties;
 import android.os.Trace;
 import android.provider.ContactsContract.Contacts;
 import android.telecom.CallState;
@@ -324,6 +325,13 @@
     private Call mConferenceLevelActiveCall = null;
 
     private boolean mIsLocallyDisconnecting = false;
+    /**
+     * Applicable for MT call, whether to show MT call instantly to user or wait for 500ms for
+     * reading direct-to-voicemail value to decide whether to route MT call to voicemail or
+     * show to user.
+     */
+    private boolean mShowCallInstantly = SystemProperties
+            .getBoolean("persist.radio.showcallinstantly", true);
 
     /**
      * Persists the specified parameters and initializes the new instance.
@@ -803,7 +811,7 @@
                 // will not need to set STATE_RINGING state prior to calling reject.
                 setState(CallState.RINGING);
                 reject(false, null);
-            } else {
+            } else if (!mShowCallInstantly) {
                 // TODO: Make this class (not CallsManager) responsible for changing
                 // the call state to STATE_RINGING.
 
@@ -864,6 +872,12 @@
             // the user or if we want to reject the call.
             mDirectToVoicemailQueryPending = true;
 
+            if (mShowCallInstantly) {
+                for (Listener l : mListeners) {
+                    Log.d(this, "show incoming call instantly");
+                    l.onSuccessfulIncomingCall(this);
+                }
+            }
             // Timeout the direct-to-voicemail lookup execution so that we dont wait too long before
             // showing the user the incoming call screen.
             mHandler.postDelayed(mDirectToVoicemailRunnable, Timeouts.getDirectToVoicemailMillis(
diff --git a/src/com/android/server/telecom/InCallController.java b/src/com/android/server/telecom/InCallController.java
index 3f9f90d..5ea583e 100755
--- a/src/com/android/server/telecom/InCallController.java
+++ b/src/com/android/server/telecom/InCallController.java
@@ -488,6 +488,10 @@
      * @param call The {@link Call}.
      */
     private void updateCall(Call call) {
+        if (call.getState() == CallState.CONNECTING) {
+            Log.d(this, "updateCall skip update for CONNECTING");
+            return;
+        }
         if (!mInCallServices.isEmpty()) {
             for (Map.Entry<ComponentName, IInCallService> entry : mInCallServices.entrySet()) {
                 ComponentName componentName = entry.getKey();
